internal/stacks/stackruntime/internal/stackeval/component.go

```diff
@@ -170,11 +159,18 @@ func (c *Component) CheckInstances(ctx context.Context, phase EvalPhase) (map[ad
 		ctx, c.instances.For(phase), c.main,
 		func(ctx context.Context) (map[addrs.InstanceKey]*ComponentInstance, tfdiags.Diagnostics) {
 			var diags tfdiags.Diagnostics
-			forEachVal := c.ForEachValue(ctx, phase)
+			forEachVal, forEachValueDiags := c.CheckForEachValue(ctx, phase)
+
+			// We can not create an instance map if the for each vaulu is not valid.
+			// We don't want to report on forEachValueDiags here because we
+			// already do this in checkValid.
+			if forEachValueDiags.HasErrors() {
```

Comment:
Presumably this shouldn't happen then? If we looked before in `checkValid`? If we would've failed earlier, having error diags means something has gone wrong in Terraform right? I'd return the `forEachValueDiags` with the other diags anyway, and an extra one saying there's an error in Terraform and please report a bug. Or, you could just panic since it's a problem we shouldn't ever see.